File: crates/async_channel/src/producer.rs
use std::future::Future;

pub struct Producer {
    pub should_stop: bool,
    pub is_running: bool,
}

impl Producer {
    pub fn new() -> Self {
        Self {
            should_stop: false,
            is_running: false,
        }
    }

    pub fn start_running(&mut self) {
        self.is_running = true;
    }

    pub fn stop(&mut self) {
        self.should_stop = true;
        self.is_running = false;
    }

    pub fn pause(&mut self) {
        self.is_running = false;
    }

    pub fn resume(&mut self) {
        // is_running is set by the caller if needed
    }

    /// No-op stub. Actual sending requires a runtime-specific queue, so
    /// concrete producers override this.
    pub fn send(&self) {
        // no-op
    }

    /// No-op stub. Push notification that new data should be sent.
    /// Concrete producers override this with actual implementation.
    pub async fn push(&self) {
        // no-op
    }

    /// No-op stub. Should be implemented for producer's non-triggered tasks.
    pub async fn start(&self) {
        // no-op
    }

    /// No-op stub. Should implement producer's non-triggered tasks.
    /// Can be used to force producer to perform tasks.
    pub async fn perform(&self) {
        // no-op
    }

    /// No-op stub. Should be implemented when producer can be modified
    /// during perform().
    pub async fn modify(&self) {
        // no-op
    }

    /// Wait until all consumers have finished processing their queues.
    /// Calls `join_fn` for each consumer index in `0..count`.
    pub async fn wait_for_processing<JF, JFut>(count: usize, mut join_fn: JF)
    where
        JF: FnMut(usize) -> JFut,
        JFut: std::future::Future<Output = ()>,
    {
        for i in 0..count {
            join_fn(i).await;
        }
    }

    /// Orchestration entry point.
    /// - Calls `register_fn` to register this producer on the channel.
    /// - If not synchronized, calls `create_task_fn` to spawn the producer task.
    pub async fn run<RF, RFut, CF>(
        &mut self,
        register_fn: RF,
        create_task_fn: CF,
        is_synchronized: bool,
    ) where
        RF: FnOnce() -> RFut,
        RFut: std::future::Future<Output = ()>,
        CF: FnOnce(),
    {
        register_fn().await;
        if !is_synchronized {
            create_task_fn();
        }
    }

    /// Mark the producer as running. Mirrors the Python `create_task` which
    /// sets `is_running = True` and spawns `start()` into an asyncio task.
    /// In pure Rust the actual task spawn is runtime-specific, so we only
    /// flip the flag here.
    pub fn create_task(&mut self) {
        self.is_running = true;
    }

    /// Send data to N consumers. Caller provides the async put operation per index.
    pub async fn send_to_all<F, Fut>(count: usize, mut send_one: F)
    where
        F: FnMut(usize) -> Fut,
        Fut: std::future::Future<Output = ()>,
    {
        for i in 0..count {
            send_one(i).await;
        }
    }

    /// Drain each consumer's queue synchronously, calling perform for each item.
    pub async fn synchronized_perform<F, Fut, JF, JFut>(
        consumers: &[SyncConsumerHandle],
        join_consumers: bool,
        timeout: f64,
        mut drain_and_perform: F,
        mut join_fn: JF,
    ) where
        F: FnMut(usize) -> Fut,
        Fut: Future<Output = ()>,
        JF: FnMut(usize, f64) -> JFut,
        JFut: Future<Output = ()>,
    {
        for (i, _handle) in consumers.iter().enumerate() {
            drain_and_perform(i).await;
            if join_consumers {
                join_fn(i, timeout).await;
            }
        }
    }

    /// Drain a queue by repeatedly checking emptiness, getting the next item,
    /// and performing on it. Stops when the queue is empty or `get` returns `None`.
    pub async fn drain_queue<T, IsEmpty, IFut, Get, GFut, Perform, PFut>(
        mut is_empty: IsEmpty,
        mut get: Get,
        mut perform: Perform,
    ) where
        IsEmpty: FnMut() -> IFut,
        IFut: Future<Output = bool>,
        Get: FnMut() -> GFut,
        GFut: Future<Output = Option<T>>,
        Perform: FnMut(T) -> PFut,
        PFut: Future<Output = ()>,
    {
        loop {
            if is_empty().await {
                break;
            }
            match get().await {
                Some(data) => perform(data).await,
                None => break,
            }
        }
    }

    pub fn is_consumers_queue_empty(consumer_priorities: &[(i64, bool)], max_priority: i64) -> bool {
        consumer_priorities
            .iter()
            .all(|(prio, empty)| *prio > max_priority || *empty)
    }
}

impl Default for Producer {
    fn default() -> Self {
        Self::new()
    }
}

/// Opaque handle for sync perform operations.
pub struct SyncConsumerHandle {
    pub index: usize,
}

File: packages/async_channel/async_channel/producer.py
#  Drakkar-Software Async-Channel
#  Copyright (c) Drakkar-Software, All rights reserved.
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 3.0 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this library.
"""
Define async_channel Producer class
"""
import asyncio
import typing

import async_channel.util.logging_util as logging

if typing.TYPE_CHECKING:
    import async_channel.channels.channel


class Producer:
    """
    A Producer is responsible for producing some output that may be placed onto the head of a queue.
    A Consumer will consume this data through the same shared queue.
    A producer doesn't need to know or care about its consumers.
    But if there is no space in the queue, it won't be able to share what it has produced.
    It manages its consumer calls by priority levels
    When the channel is synchronized priority levels are used to priorities or delay consumer calls
    """

    def __init__(self, channel: "async_channel.channels.channel.Channel"):
        self.logger = logging.get_logger(self.__class__.__name__)

        # Related async_channel instance
        self.channel: "async_channel.channels.channel.Channel" = channel

        """
        Should only be used with .cancel()
        """
        self.produce_task: typing.Optional[asyncio.Task] = None

        """
        Should be used as the perform while loop condition
            while(self.should_stop):
                ...
        """
        self.should_stop: bool = False

        """
        Should be used to know if the producer is already started
        """
        self.is_running: bool = False

    async def send(self, data: typing.Any) -> None:
        """
        Send to each consumer data though its queue
        :param data: data to be put into consumers queues

        The implementation should use 'self.async_channel.get_consumers'
        Example
            >>> for consumer in self.async_channel.get_consumers():
            >>>     await consumer.queue.put({
            >>>         "my_key": my_value
            >>>     })
        """
        for consumer in self.channel.get_consumers():
            await consumer.queue.put(data)

    async def push(self, **kwargs) -> None:
        """
        Push notification that new data should be sent implementation
        When nothing should be done on data : self.send()
        """

    async def start(self) -> None:
        """
        Should be implemented for producer's non-triggered tasks
        """

    async def pause(self) -> None:
        """
        Called when the channel runs out of consumer
        """
        self.logger.debug("Pausing...")
        self.is_running = False
        # Triggers itself if not already paused
        if not self.channel.is_paused:
            self.channel.is_paused = True

    async def resume(self) -> None:
        """
        Called when the channel is no longer out of consumer
        """
        self.logger.debug("Resuming...")
        # Triggers itself if not already resumed
        if self.channel.is_paused:
            self.channel.is_paused = False

    async def perform(self, **kwargs) -> None:
        """
        Should implement producer's non-triggered tasks
        Can be use to force producer to perform tasks
        """

    async def modify(self, **kwargs) -> None:
        """
        Should be implemented when producer can be modified during perform()
        """

    async def wait_for_processing(self) -> None:
        """
        Should be used only with SupervisedConsumers
        It will wait until all consumers have notified that their consume() method have ended
        """
        await asyncio.gather(
            *(consumer.join_queue() for consumer in self.channel.get_consumers())
        )

    async def synchronized_perform_consumers_queue(
        self, priority_level: int, join_consumers: bool, timeout: float
    ) -> None:
        """
        Empties the queue synchronously for each consumers
        :param priority_level: the consumer minimal priority level
        :param join_consumers: True if consumer tasks should be joined. Avoids orphaned tasks to run without
        :param timeout: Time to wait for consumers in join call
        waiting for them when started before this check (when check, their queue is empty but a task is running)
        """
        for consumer in self.channel.get_prioritized_consumers(priority_level):
            while not consumer.queue.empty():
                await consumer.perform(await consumer.queue.get())
            if join_consumers:
                await consumer.join(timeout)

    async def stop(self) -> None:
        """
        Stops non-triggered tasks management
        """
        self.should_stop = True
        self.is_running = False
        if self.produce_task:
            self.produce_task.cancel()

    def create_task(self) -> None:
        """
        Creates a new asyncio task that contains start() execution
        """
        self.is_running = True
        self.produce_task = asyncio.create_task(self.start())

    async def run(self) -> None:
        """
        Start the producer main task
        Shouldn't start the producer main task if the channel is synchronized
        Should always call
        >>> self.async_channel.register_producer
        """
        await self.channel.register_producer(self)
        if not self.channel.is_synchronized:
            self.create_task()

    def is_consumers_queue_empty(self, priority_level: int) -> bool:
        """
        Check if consumers queue are empty
        :param priority_level: the consumer minimal priority level
        :return: the check result
        """
        for consumer in self.channel.get_consumers():
            if consumer.priority_level <= priority_level and not consumer.queue.empty():
                return False
        return True
